packages/contracts/contracts/FlippenZap.sol

```diff
@@ -0,0 +1,248 @@
+    /// @param _collToken The address of collateral token used in eBTC system
+    /// @param _ebtcToken The address of eBTC token
+    /// @param _balancerVault The address of Balancer Vault
+    /// @param _balancerPoolId The pool ID of eBTC-stETH pair in Balancer Vault
+    constructor(
+        address _borrowerOperations,
+        address _priceFeed,
+        address _cdpManager,
+        address _collToken,
+        address _ebtcToken,
+        address _balancerVault,
+        bytes32 _balancerPoolId
+    ) public {
+        require(_borrowerOperations != address(0), "FlippenZap: empty BorrowerOperations address");
+        require(_priceFeed != address(0), "FlippenZap: empty PriceFeed address");
+        require(_cdpManager != address(0), "FlippenZap: empty CdpManager address");
+        require(_collToken != address(0), "FlippenZap: empty collateral token address");
+        require(_ebtcToken != address(0), "FlippenZap: empty eBTC token address");
+        require(_balancerVault != address(0), "FlippenZap: empty Balancer vault address");
+        require(_balancerPoolId != bytes32(0), "FlippenZap: empty Balancer pool ID");
+
+        ebtcBorrowerOperations = IBorrowerOperations(_borrowerOperations);
+        ebtcPriceFeed = IPriceFeed(_priceFeed);
+        ebtcCdpManager = ICdpManager(_cdpManager);
+        ebtcCollToken = ICollateralToken(_collToken);
+        ebtcToken = IEBTCToken(_ebtcToken);
+        balancerVault = IBalancerV2Vault(_balancerVault);
+        balancerPoolId = _balancerPoolId;
+
+        ebtcCollToken.approve(_borrowerOperations, type(uint256).max);
+        ebtcToken.approve(_borrowerOperations, type(uint256).max);
+
+        ebtcCollToken.approve(_balancerVault, type(uint256).max);
+        ebtcToken.approve(_balancerVault, type(uint256).max);
+    }
+
+    /// @notice This function allows caller to create a leveraged CDP with an initial collateral.
+    /// @notice The caller has to approve this contract as a valid PositionManager via BorrowerOperations
+    /// @notice and approve this contract to spend _initialStETH amount of collateral
+    /// @param _initialStETH The initial amount of stETH used as collateral
+    /// @param _leverage The expected leverage for this long position, should be no more than MAX_REASONABLE_LEVERAGE
+    /// @return The CdpId created with owner as the caller
+    function enterLongEth(uint256 _initialStETH, uint256 _leverage) external returns (bytes32) {
+        require(_initialStETH > MIN_NET_STETH_BALANCE, "FlippenZap: initial principal too small");
+        require(_leverage <= MAX_REASONABLE_LEVERAGE, "FlippenZap: _leverage is too big");
+        require(_leverage > 0, "FlippenZap: _leverage is too small");
+
+        // under the covers:
+        // mints eBTC against stETH collateral and
+        // sells the eBTC back for stETH.
+
+        bytes32 _newCdp;
+        uint256 _collToLoan = _leverage > 1 ? (_leverage - 1) * _initialStETH : 0;
```

Comment:
I am not sure this calculation is correct. 

Maybe it is.

I always get confused about the role of the gas stipend in some calculations, so I'm just thinking out loud:

Assume the user has 10 stETH and the gas stipend is 1 stETH, for the sake of simplicity. Then the user wants to leverage 5x. This means they effectively have 9 stETH at their disposal. Shouldn't this be the number multiplied by 5x, i.e., 45 stETH as a result? Not 50 stETH?

Reply:
Yep, the value of liquidator stipend should be reduced from initial principal before the calculation for flashloan. Fixed.